File: MK-52_ESP32/src/SD_Manager.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include <SD.h>
#include "SD_Manager.hpp"
#include "UniversalValue.hpp"

//#define __DEBUG

// const char SD_Error_NotMounted[] PROGMEM = "Error: Not mounted";
// const char SD_Error_FileNameTooLong[] PROGMEM = "Err: Long name";
// const char SD_Error_FileNotFound[] PROGMEM = "Err: Not found";
// const char SD_Error_WriteFailed[] PROGMEM = "Err: SD fail";
// const char SD_Error_FileAccessError[] PROGMEM = "Err: Write fail";
// const char SD_Error_OutOfMemory[] PROGMEM = "Err: Out of memory";
// const char SD_Error_DeleteFailed[] PROGMEM = "Err: Delete fail";

const char SD_root[] PROGMEM = "/";

using namespace MK52_Hardware;

//
// Init and status update
//
void SD_Manager::init(){
    SDMounted = SD.begin();
    _buffer = (char *)malloc( DIRECTORY_LIST_SIZE);
    _current_Dir_Name = (char *)malloc( CURRENT_DIR_LEN + CURRENT_FILE_LEN * 2 + 3);
    #ifdef __DEBUG
    if( _buffer == NULL || _current_Dir_Name == NULL){
        Serial.println("File List malloc busted!");
        return;
    }
    #endif
    _clearItems();
    memset( _current_Dir_Name, 0, CURRENT_DIR_LEN + CURRENT_FILE_LEN * 2 + 3);
    _current_File_Name = _current_Dir_Name + CURRENT_DIR_LEN + 1;
    _text = _current_File_Name + CURRENT_FILE_LEN + 1;
    setFolder_P(SD_root);
}

void SD_Manager::checkRootExists(){
    if(!SDMounted) return;
    File root = SD.open( _current_Dir_Name);
    if(!root){
        _resetRoot();
        return;
    }
    if(!root.isDirectory()) _resetRoot();
    root.close();
}

bool SD_Manager::checkEntityExists( const char *name){
    if( !SDMounted ) return false;
    return SD.exists(name);
}

void SD_Manager::nextListingPosition(){
    if( _nItems <= 0){
        listingPosition = -1;
        return;
    }
    listingPosition++;
    if(listingPosition>=_nItems) listingPosition=_nItems-1;
}

void SD_Manager::previousListingPosition(){
    if( _nItems <= 0){
        listingPosition = -1;
        return;
    }
    listingPosition--;
    if(listingPosition<0) listingPosition=0;
}

//
// Lists directory and displays selector
//
void SD_Manager::getFolderListing( char *Lines[], uint8_t nLines, uint8_t lineLen, char *name){
    if( SDMounted)
        strcpy( Lines[0], getFolderNameTruncated( SCREEN_COLS-1));
    else
        strcpy_P( Lines[0], PSTR("Insert SD card and reboot"));
    int startLine = 0;
    int stopLine = nLines-2;
    if( listingPosition > stopLine){
        startLine = listingPosition-stopLine;
        stopLine += startLine;
    }
    for( uint8_t i=1; i<nLines; i++, startLine++){
        char *Ln = Lines[i]; 
        memset( Ln+1, 0, lineLen);
        if( startLine > stopLine) continue;
        Ln[ 0] = (startLine == listingPosition)? (char)172 : ' ';
        Ln[ 1] = ' ';
        strncpy( Ln+2, getItemString(startLine), lineLen-2);
    }
}

//
// gets an item from selector
//
char *SD_Manager::getItemFromListing(){
    memset( _current_File_Name, 0, CURRENT_FILE_LEN);
    if( listingPosition<0) return _current_File_Name;
    int16_t loc = *(getItemPtr( listingPosition));
    if( loc < 0 || loc >= _nItems) return _current_File_Name;
    File current_Dir = _getCurrentDir();
    if( !current_Dir) return _current_File_Name;
    File file = current_Dir.openNextFile();
    while( loc>0 && file){ 
        file = current_Dir.openNextFile();
        loc--;
    }
    current_Dir.close();
    if( !file) return _current_File_Name;
    strncpy( _current_File_Name, file.name(), CURRENT_FILE_LEN-1);
    _current_File_Name[CURRENT_FILE_LEN-1] = 0;
    file.close();
    #ifdef __DEBUG
    Serial.print("Selected: ");
    Serial.println(_current_File_Name);
    #endif
    return _current_File_Name;
}

bool SD_Manager::deleteEntity( const char *name){
    if( !SDMounted) return true;
    if( strlen(name) == 1) return false;
    #ifdef __DEBUG
    Serial.print( "Deleting: ");
    Serial.println( name);
    #endif
    if( !SD.exists( name)) return false;
    File root = SD.open( name);
    bool dir = root.isDirectory();
    if( dir){ // delete recursively
        File file = root.openNextFile();
        while(file){
        if( deleteEntity( file.name())) break;
        file = root.openNextFile();
        } // while
    }//if
    root.close();
    return !(dir? SD.rmdir(name): SD.remove( name));
}

char *SD_Manager::makeEntityName( char *name){
    memset( _text, 0, CURRENT_FILE_LEN);
    if(strlen(name) <= 0) return _text; // no name given
    strncpy( _text, _current_Dir_Name, CURRENT_FILE_LEN-3);
    int ln = strlen(_text);
    if( ln>1) _text[ln++] = '/'; // add slash if not at root
    strncpy( _text+ln, name, CURRENT_FILE_LEN-2-ln);
    return _text;
}

//
// returns true if an error
//
bool SD_Manager::openFile( char *path, bool write){
    if( !SDMounted) return true;
    closeFile();
    if( path == NULL){
        if( isAtFolder()) return true;
        path = getItemFromListing();
    }
    if( strlen( path) <= 1 ) return true; // cannot overwrite root!
    if( write)
        _current_File = SD.open( (const char*)path, FILE_WRITE);
    else
        _current_File = SD.open( (const char*)path);
    if( _current_File){
        _current_File_open = true;
        #ifdef __DEBUG
        Serial.print("Opened ");
        Serial.print( path);
        Serial.println( write? " for writing": " for reading");
        #endif
        return false;
    }
    #ifdef __DEBUG
    else{
        Serial.print("Failed to open ");
        Serial.println( path);
    }
    #endif
    return true;
}

bool SD_Manager::openFile_P( const char *path, bool write){
    strncpy_P( _current_File_Name, path, CURRENT_FILE_LEN);
    _current_File_Name[CURRENT_FILE_LEN] = 0;
    return openFile( _current_File_Name, write);
}

void SD_Manager::closeFile(){
    if( !SDMounted) return;
    if( _current_File_open) _current_File.close();
    _current_File_open = false;
}

bool SD_Manager::print( char *message){
    if( !SDMounted) return true;
    if( !_current_File_open) return true;
    if( !_current_File.print((const char*)message)) return true;
    return false;
}

bool SD_Manager::print_P( const char *message){
    strncpy_P( _text, message, CURRENT_FILE_LEN-1);
    _text[CURRENT_FILE_LEN-1] = 0;
    return( print( _text));
}

bool SD_Manager::println( char *message){
    if( !SDMounted) return true;
    if( !_current_File_open) return true;
    if( !_current_File.println((const char*)message)) return true;
    return false;
}

bool SD_Manager::println_P( const char *message){
    strncpy_P( _text, message, CURRENT_FILE_LEN-1);
    _text[CURRENT_FILE_LEN-1] = 0;
    return( println( _text));
}

bool SD_Manager::read( char *buffer, int32_t n){
    if( !SDMounted) return true;
    if( !_current_File_open) return true;
    while( n>0 && _current_File.available()){
        *(buffer++) = (char) _current_File.read();
        n--;
    }
    *buffer = 0; // safety zero
    return n>0; // not all symbols read - file end
}

bool SD_Manager::readln( char *buffer, int32_t n){
    if( !SDMounted) return true;
    if( !_current_File_open) return true;
    uint8_t b = 0;
    while( _current_File.available()){
        uint8_t b = _current_File.read();
        if( b ==_CR_) continue;
        if( b ==_LF_) break;
        if(n>0){
            *(buffer++) = (char)b;
            n--;
        }
    }
    *buffer = 0; // safety zero
    return  !_current_File.available();
}

void SD_Manager::createFolder( char * name){
    if( !SDMounted) return;
    if(strlen(name) <= 0) return; // no name given
    makeEntityName( name);
    #ifdef __DEBUG
    Serial.print( "Creating folder: ");
    Serial.println( _text);
    #endif
    if( !SD.mkdir(_text)) return;
    setFolder( _text);
}

void SD_Manager::upFolder(){
    if( !SDMounted) return;
    int ln = strlen(_current_Dir_Name);
    if( ln <= 1) return; // already at root
    while( ln>0 && _current_Dir_Name[ln] != '/') ln--;
    strcpy( _current_File_Name, _current_Dir_Name+ln+1);
    _current_Dir_Name[ln] = 0;
    #ifdef __DEBUG
    Serial.print("Going up: ");
    Serial.println( _current_Dir_Name);
    Serial.print( "Proposed location: ");
    Serial.println( _current_File_Name);
    #endif
    readFolderItems( _current_File_Name);
}

//
// returns true if a run is required
//
bool SD_Manager::stepIn(char *name){
    if( !SDMounted) return false;
    File tmp = SD.open( (const char*)name);
    if(!tmp) return false;
    if( tmp.isDirectory()){
        tmp.close();
        setFolder( name);
        return false;
    }
    tmp.close();
    return true;
}

void SD_Manager::setFolder( char *name){
    strncpy( _current_Dir_Name, name, CURRENT_DIR_LEN-1);
    _current_Dir_Name[CURRENT_DIR_LEN-1] = 0;
    readFolderItems();
}

void SD_Manager::setFolder_P( const char *name){
    strncpy_P( _current_Dir_Name, name, CURRENT_DIR_LEN-1);
    _current_Dir_Name[CURRENT_DIR_LEN-1] = 0;
    readFolderItems();
}

char *SD_Manager::getFolderNameTruncated( int8_t n){
    int ln = strlen(_current_Dir_Name);
    if( ln > n-1) ln = ln - n + 1;
    else ln = 0;
    strcpy( _text, _current_Dir_Name+ln);
    ln = strlen(_text);
    if( ln>1){
        _text[ln++] = '/';
        _text[ln] = 0;
    }
    return _text;
}

void SD_Manager::readFolderItems(char *location){
    if( !SDMounted) return;
    _clearItems();
    if( !checkEntityExists(_current_Dir_Name)) return;
    File current_Dir = _getCurrentDir();
    if(!current_Dir) return;
    File file = current_Dir.openNextFile();
    int16_t counter = 0;
    int16_t index = 0;
    int16_t pos2 = -1;
    while( file){
        char *name = (char *)_formEntityName( file);
        if( MK52_Interpreter::UniversalValue::_startsWith( name, location))
            pos2 = counter;
        bool isdir = file.isDirectory();
        index = _locateAlphabetic( name, isdir);
        #ifdef __DEBUG
        Serial.print(name);
        Serial.print(" is located at ");
        Serial.println(index);
        #endif
        if( _insertItem( name, counter++, index, isdir)) break; // too many items
        file = current_Dir.openNextFile();
    }
    current_Dir.close();
    if( _nItems > 0) listingPosition = 0;
    if( location == NULL) return;
    Serial.print("Looking for location: ");
    Serial.println(location);
    for( int i=0; i<_nDirs; i++){
        if( *(getItemPtr(i)) != pos2) continue;
        listingPosition = i;
        break;
    }
}

void SD_Manager::_resetRoot(){
    #ifdef __DEBUG
    Serial.println("Reset to root");
    #endif
    strncpy_P( _current_Dir_Name, SD_root, CURRENT_DIR_LEN);
}

File SD_Manager::_getCurrentDir(){
    checkRootExists();
    return SD.open( _current_Dir_Name);
}

void SD_Manager::_clearItems(){
    _nDirs = 0;
    _nItems = 0;
    listingPosition = -1;
    memset( _buffer, 0, DIRECTORY_LIST_SIZE);
}

//
// Scans items and locates the added item alphabetically:
// Directories are sorted first
// Returns the slot number for insertion
//
int16_t SD_Manager::_locateAlphabetic(const char *name, bool isDir){
    if( _nItems <= 0) return 0;
    int16_t slot = 0;
    char *ptr = (char *)name;
    if( isDir){
        while( slot < _nDirs){
            if( strncmp( ptr, getItemString(slot), 21) <= 0)
                return slot;
            slot++;
        }
        return _nDirs;
    }
    else{
        slot = _nDirs;
    }
    while( slot < _nItems){
        if( strncmp( ptr, getItemString(slot), 21) <= 0)
            return slot;
        slot++;
    }
    return _nItems;
}

//
// Returns true if the maximum directory length is exceeded
//
bool SD_Manager::_insertItem(const char *name, int16_t pos, int16_t slot, bool isDir){
    if( slot>=DIRECTORY_LIST_NITEMS-1) return true;
    if( slot>=_nItems) slot = _nItems;
    int16_t toMove = _nItems - slot;
    char *ptr = _buffer + slot * SCREEN_COLS;
    if( toMove > 0) memmove( ptr+SCREEN_COLS, ptr, toMove*SCREEN_COLS);
    int16_t toCopy = strlen( name);
    if( toCopy > SCREEN_COLS-3) toCopy = SCREEN_COLS-3;
    memcpy( ptr, &pos, 2);
    memset( ptr+2, 0, SCREEN_COLS-2);
    memcpy( ptr+2, name, toCopy);
    if( isDir) _nDirs++;
    _nItems++;
    return false;
}

char *SD_Manager::_stripFolders( const char *name){
  int16_t lastHash = strlen( (char*) name) - 1;
  while( lastHash>0 && name[lastHash] != '/') lastHash--;
  return (char *)name+lastHash+1; 
}

char *SD_Manager::_formEntityName( File f){
    memset( _text, 0, SCREEN_COLS-3);
    const char *name = f.name();
    char *namePtr = _stripFolders( name);
    uint8_t nameLen = strlen( namePtr);
    if( nameLen > SCREEN_COLS - 9) nameLen = SCREEN_COLS - 9;
    if( nameLen>0) memcpy( _text, namePtr, nameLen);
    for( uint16_t i=strlen(_text); i<SCREEN_COLS-9; i++) _text[i] = ' ';
    return _appendFileSize(f);
}

char *SD_Manager::_appendFileSize( File f){
    char *str = _text + SCREEN_COLS - 9;
    if( f.isDirectory()){
        strcpy_P( str, PSTR(" [DIR]"));
        return _text;
    }
    size_t size = f.size();
    if( size < 8192){
        sprintf_P( str, PSTR(" %04db"), (int)size);
        return _text;
    }
    size >>= 10;
    if( size < 8192){
        sprintf_P( str, PSTR(" %04dk"), (int)size);
        return _text;
    }
    size >>= 10;
    if( size < 8192){
        sprintf_P( str, PSTR(" %04dM"), (int)size);
        return _text;
    }
    size >>= 10;
    sprintf_P( str, PSTR(" %04dG"), (int)size);
}

// //
// // Locates an existing BASIC file on disk
// //
// bool SD_Manager::_locateBASICFile( const char *name){
//   size_t len = strlen( name);
//   if( _nameLengthCheckMantra( len)) return true;

//   // first, try the absolute name; no memory allocation needed
//   if( SD.exists(name)){
//     strncat2( _vars->currentFile, name, CURRENT_FILE_LEN);
//     return false;
//   }
  
//   // next, try directory + name
//   len += strlen( _vars->currentDir) + 5;
//   if( _nameLengthCheckMantra( len)) return true;
//   char *tmpName = (char *) malloc( len);
//   if( tmpName == NULL){
//     LastError = SD_Error_OutOfMemory;
//     return true;
//   }
//   strcpy( tmpName, _vars->currentDir);
//   while( *name == '/') name++;
//   strcat( tmpName, name);
//   if( _lookForFileMantra2( tmpName)) return false; // note it frees tmpName

//   // next, try directory + name + ".bas"
//   size_t eos = strlen( tmpName);
//   strcat( tmpName, SD_DefaultExt1);
//   if( _lookForFileMantra2( tmpName)) return false; // note it frees tmpName

//   // finally, try directory + name + ".BAS"
//   tmpName[eos] = _NUL_;
//   strcat( tmpName, SD_DefaultExt2);
//   if( _lookForFileMantra2( tmpName)) return false; // note it frees tmpName

//   // if we are here, nobody freed the memory
//   free( tmpName);
//   LastError = SD_Error_FileNotFound;
//   return true;
// }

// bool SD_Manager::_lookForFileMantra1( char *tmpName){
//   #ifdef __DEBUG
//   Serial.print( "Looking for file: ");
//   Serial.println( tmpName);
//   #endif
//   if( !SD.exists(tmpName)) return false;
//   _currentFile = SD.open( tmpName);
//   free( tmpName);
//   if( !_currentFile) LastError = NULL;
//   return true;
// }
// bool SD_Manager::_lookForFileMantra2( char *tmpName){
//   #ifdef __DEBUG
//   Serial.print( "Looking for file: ");
//   Serial.println( tmpName);
//   #endif
//   if( !SD.exists(tmpName)) return false;
//   strncat2( _vars->currentFile, tmpName, CURRENT_FILE_LEN);
//   free( tmpName);
//   return true;
// }
// bool SD_Manager::_formFileMantra( const char *name, char *dest){
//   if( name == NULL) return false;
//   int16_t len = strlen( name);
//   if( len<2){
//     LastError = SD_Error_FileAccessError;
//     return true;
//   }
//   if( name[0] != '/') len += strlen( _vars->currentDir) + 1;
//   if( _nameLengthCheckMantra( len)) return true;
//   dest[0] = _NUL_;
//   if( name[0] != '/') strcat( dest, _vars->currentDir);
//   strcat( dest, name);
//   return false;
// }

//
// Checks if directory structure exists, if not - creates
//
// bool SD_Manager::_checkFolderStructure( char *name){
//   if( SD.exists( name)) return false;
//   int16_t lastHash = strlen( name) - 1;
//   while( lastHash>0 && name[lastHash] != '/') lastHash--;
//   if( lastHash <= 0) return false; 
//   for( int i=1; i<=lastHash; i++){
//     if( name[i] != '/') continue;
//     char tmp = name[i];
//     name[i] = _NUL_;
//     if( SD.exists( (const char*)name)){
//       #ifdef __DEBUG
//       Serial.print(name);
//       Serial.println(" is OK");
//       #endif
//       name[i] = tmp;
//       continue;
//     }
//     #ifdef __DEBUG
//     Serial.print("Creating ");
//     Serial.println(name);
//     #endif
//     bool ret = SD.mkdir( (const char*)name);
//     name[i] = tmp;
//     if( !ret){
//       #ifdef __DEBUG
//       Serial.println("Create failed, exiting");
//       #endif
//       LastError = SD_Error_FileAccessError;
//       return true;
//     }
//   }
//   // directory structure checked
//   return false;
// }

File: MK-52_Simulator/SD_Manager.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    public class SD_Manager
    {
        private Dictionary<string, DirectoryInfo> _directoryListing
            = new Dictionary<string, DirectoryInfo>();
        private Dictionary<string, FileInfo> _fileListing
            = new Dictionary<string, FileInfo>();
        private List<string> _itemListing = new List<string>();

        private string _current_Dir_Name = "C:/";
        private string _current_File_Name = "";
        private DateTime _lastDirectoryRead = DateTime.MinValue; // Windows-only

        public string StatusFileName = "";

        public const byte _LF_ = 10;
        public const byte _CR_ = 13;

        public bool SDMounted = false; // fake for simulator only
        public UInt64 cardSize = 4000000000L; // fake for simulator only
        public int listingPosition = -1;
        public int _nDirs = 0;
        public int _nItems = 0;

        public bool _current_File_open = false;
        public StreamReader __sr = null;
        public StreamWriter __sw = null;
        public FileStream __fs = null;
        public string __text = "";
        
        public void init()
        {
            _clearItems();
            _current_Dir_Name = Directory.GetParent(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)).FullName;
            //if ( Environment.OSVersion.Version.Major >= 6 )
            //    _current_Dir_Name = Directory.GetParent(_current_Dir_Name).ToString();
            SDMounted = Directory.Exists(_current_Dir_Name);
            if (!SDMounted) return;
            _current_Dir_Name += "\\MK52Simulator";
            if (!Directory.Exists(_current_Dir_Name))
                Directory.CreateDirectory(_current_Dir_Name);
            SDMounted = Directory.Exists(_current_Dir_Name);
            StatusFileName = _current_Dir_Name + "\\_MK52_Status.DAT";
            _current_Dir_Name = _current_Dir_Name.Replace('\\', '/');
            setFolder_P(_current_Dir_Name);
        }

        public string getFolderName()
        {
            return _current_Dir_Name;
        }

        public string getWindowsFolderName()
        {
            return _current_Dir_Name.Replace('/', '\\');
        }

        public string getWindowsFileName()
        {
            return _current_File_Name.Replace('/', '\\');
        }

        public string getFolderNameTruncated( int n)
        {
            int ln = _current_Dir_Name.Length;
            if( ln >= n-1)
            {
                return _current_Dir_Name.Substring(ln - n + 1) + "/";
            }
            return _current_Dir_Name + "/";
        }

        public void nextListingPosition()
        {
            if (_nItems <= 0)
            {
                listingPosition = -1;
                return;
            }
            listingPosition++;
            if (listingPosition >= _nItems) listingPosition = _nItems - 1;
        }

        public void previousListingPosition()
        {
            if (_nItems <= 0)
            {
                listingPosition = -1;
                return;
            }
            listingPosition--;
            if (listingPosition < 0) listingPosition = 0;
        }
        
        public bool isAtFolder()
        {
            return listingPosition <_nDirs;
        }

        /// <summary>
        /// Gets an item from selector
        /// This is Windows-only implementation
        /// </summary>
        /// <returns></returns>
        public string getItemFromListing()
        {
            _current_File_Name = "";
            if( listingPosition<0 || listingPosition >= _nItems) return _current_File_Name;
            string searchKey = _itemListing[listingPosition];
            if (_directoryListing.ContainsKey(searchKey))
                _current_File_Name = _directoryListing[searchKey].FullName;
            if (_fileListing.ContainsKey(searchKey))
                _current_File_Name = _fileListing[searchKey].FullName; 
            return _current_File_Name;
        }

        //abstract public void checkRootExists();

        public bool checkEntityExists(string name)
        {
            name = name.Replace( '/', '\\');
            return File.Exists(name) || Directory.Exists(name);
        }
        
        public string makeEntityName( string name)
        {
            StringBuilder sb = new StringBuilder();
            if (name.Length <= 0) return sb.ToString(); // no name given
            sb.Append(_current_Dir_Name);
            sb.Append('/');
            sb.Append(name);
            return sb.ToString();
        }

        public string getItemString(int n)
        {
            if (n >= _itemListing.Count) return "";
            return _itemListing[n];
        }

        public void setFolder(string name)
        {
            setFolder_P(name);
        }

        public void setFolder_P(string name)
        {
            _current_Dir_Name = name.Replace('\\', '/');
            readFolderItems();
        }

        public void readFolderItems()
        {
            readFolderItems("", true);
        }

        public void readFolderItems(string location, bool resetListingPosition)
        {
            if( !SDMounted) return;
            _clearItems();
            if (resetListingPosition) listingPosition = -1;
            if( !checkEntityExists(_current_Dir_Name)) return;
            string wdn = getWindowsFolderName();
            if (!Directory.Exists(wdn)) return;
            DirectoryInfo root = new DirectoryInfo(wdn);
            DirectoryInfo[] dis = root.GetDirectories();
            foreach (DirectoryInfo di in dis)
            {
                string en = _formEntityName(di);
                if( _directoryListing.ContainsKey( en)) continue;
                _directoryListing[en] = di;
            }
            FileInfo[] fis = root.GetFiles();
            foreach (FileInfo fi in fis)
            {
                string en = _formEntityName(fi);
                if (_fileListing.ContainsKey(en)) continue;
                _fileListing[en] = fi;
            }
            List<string> tmp = new List<string>();
            foreach (string s in _directoryListing.Keys)tmp.Add(s);
            tmp.Sort();
            foreach (string s in tmp) _itemListing.Add(s);
            _nDirs = _itemListing.Count;
            tmp.Clear();
            foreach (string s in _fileListing.Keys) tmp.Add(s);
            tmp.Sort();
            foreach (string s in tmp) _itemListing.Add(s);
            _nItems = _itemListing.Count;
            if (location.Length == 0)
            {
                if (listingPosition < 0 && _nItems > 0) listingPosition = 0;
                if (listingPosition >= _nItems) listingPosition = _nItems - 1; 
                return;
            }
            for (int i = 0; i < _nDirs; i++)
            {
                if (!_itemListing[i].StartsWith(location)) continue;
                listingPosition = i;
                break;
            }
        }

        /// <summary>
        /// In Windows, we re-read the folder once in a while, as other processes may add files/folders
        /// </summary>
        /// <param name="Lines"></param>
        /// <param name="nLines"></param>
        /// <param name="lineLen"></param>
        /// <returns></returns>
        public string[] getFolderListing(string[] Lines, int nLines, int lineLen)
        {
            if (DateTime.Now - _lastDirectoryRead > TimeSpan.FromSeconds(10.0))
            {
                readFolderItems("", false);
                _lastDirectoryRead = DateTime.Now;
            }
            return getFolderListing( Lines, nLines, lineLen, null);
        }

        public string[] getFolderListing(string[] Lines, int nLines, int lineLen, string name)
        {
            if( SDMounted)
                Lines[0] = getFolderNameTruncated( LCD_Manager.SCREEN_COLS-1);
            else
                Lines[0] = "Insert SD card and reboot";
            int startLine = 0;
            int stopLine = nLines - 2;
            if (listingPosition > stopLine)
            {
                startLine = listingPosition - stopLine;
                stopLine += startLine;
            }
            for ( int i = 1; i < nLines; i++, startLine++)
            {
                Lines[i] = (startLine == listingPosition) ? "\u00ac ": "  ";
                Lines[i] += getItemString(startLine);
            }
            return Lines;
        }

        public void setListingPosition( int pos)
        {
            listingPosition = ( pos>=_nItems)? _nItems-1: pos;
        }

        /// <summary>
        /// Simulator-only (ESP32 is recursive)
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public bool deleteEntity(string name)
        {
            if( !SDMounted) return true;
            if( name.Length < 3) return false;
            if (Directory.Exists(name))
            {
                try
                {
                    Directory.Delete(name, true);
                    return false;
                }
                catch
                {
                    return true;
                }
            }
            if (File.Exists(name))
            {
                try
                {
                    File.Delete(name);
                    return false;
                }
                catch
                {
                    return true;
                }
            }
            return false;
        }

        public void createFolder(string name)
        {
            if( !SDMounted) return;
            if(name.Length <= 0) return; // no name given
            __text = makeEntityName( name);
            string tmp = __text.Replace('/', '\\');
            try
            {
                if( !Directory.Exists( tmp)) Directory.CreateDirectory(tmp);
                setFolder( __text);
            }
            catch
            {
            }
        }

        public void upFolder()
        {
            if( !SDMounted) return;
            if (_current_Dir_Name.Length <= 2) return; // already at root
            int index = _current_Dir_Name.LastIndexOf('/');
            if (index < 2) return;
            _current_File_Name = _current_Dir_Name.Substring(index+1);
            _current_Dir_Name = _current_Dir_Name.Substring(0, index);
            readFolderItems( _current_File_Name, true);
        }

        /// <summary>
        /// Steps into a folder or executes a file
        /// </summary>
        /// <param name="name"></param>
        /// <returns>returns true if a run is required</returns>
        public bool stepIn(string name)
        {
            if( !SDMounted) return false;
            if( Directory.Exists( name.Replace('/', '\\')))
            {
                setFolder( name);
                return false;
            }
            return true;
        }

        /// <summary>
        /// Simulates ESP32 SD utilities
        /// </summary>
        /// <param name="path">Fake "Linux" file name</param>
        /// <param name="write">true to write</param>
        /// <returns>True if an error</returns>
        public bool openFile(string path, bool write)
        {
            return openFile_P(path, write);
        }

        /// <summary>
        /// Simulates ESP32 SD utilities
        /// </summary>
        /// <param name="path">Windows file name</param>
        /// <param name="write">true to write</param>
        /// <returns>True if an error</returns>
        public bool openFile_P(string path, bool write)
        {
            if( !SDMounted) return true;
            closeFile();
            if( path.Length == 0){
                if( isAtFolder()) return true;
                path = getItemFromListing();
            }
            if( path.Length < 4 ) return true; // cannot overwrite root, e.g. C:\
            try
            {
                if (write)
                {
                    __fs = File.Open(path, FileMode.Create, FileAccess.Write, FileShare.Read);
                    __sw = new StreamWriter(__fs);
                }
                else
                {
                    __fs = File.Open(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
                    __sr = new StreamReader(__fs);
                }
                _current_File_Name = path.Replace('\\', '/');
                _current_File_open = true;
                return false;
            }
            catch
            {
                closeFile();
            }
            return true;
        }

        public void closeFile()
        {
            if( __sr != null)
            {
                __sr.Close();
                __sr = null;
            }
            if( __sw != null)
            {
                __sw.Close();
                __sw = null;
            }
            if( __fs != null)
            {
                __fs.Close();
                __fs = null;
            }
            _current_File_open = false;
        }

        public bool print(string message)
        {
            if( !SDMounted) return true;
            if( !_current_File_open) return true;
            try
            {
                __sw.Write(message);
            }
            catch
            {
                return true;
            }
            return false;
        }

        public bool print_P(string message)
        {
            return print(message);
        }

        public bool println(string message)
        {
            if( !SDMounted) return true;
            if( !_current_File_open) return true;
            try
            {
                __sw.WriteLine( message);
            }
            catch
            {
                return true;
            }
            return false;
        }

        public bool println_P(string message)
        {
            return println( message);
        }

        public bool read(string buffer, int n)
        {
            __text = "";
            if (!SDMounted) return true;
            if (!_current_File_open) return true;
            StringBuilder sb = new StringBuilder();
            while (n > 0 && !__sr.EndOfStream)
            {
                sb.Append((char)__sr.Read());
                n--;
            }
            __text = sb.ToString();
            return n > 0; // not all symbols read - file end
        }

        public bool readln( string buffer, int n)
        {
            __text = "";
            if (!SDMounted) return true;
            if (!_current_File_open) return true;
            byte b = 0;
            StringBuilder sb = new StringBuilder();
            while (!__sr.EndOfStream)
            {
                b = (byte)__sr.Read(); // to match ESP32 method
                if (b == _CR_) continue;
                if (b == _LF_) break;
                if (n > 0)
                {
                    sb.Append((char)b);
                    n--;
                }
            }
            __text = sb.ToString();
            return __sr.EndOfStream;
        }

        private void _clearItems()
        {
            _nDirs = 0;
            _nItems = 0;
            _directoryListing .Clear();
            _fileListing.Clear();
            _itemListing.Clear();
        }

        /// <summary>
        /// C++ dummy
        /// </summary>
        private int _locateAlphabetic(string name, bool isDir)
        {
            return 0;
        }

        /// <summary>
        /// C++ dummy
        /// </summary>
        private bool _insertItem(string name, int pos, int slot, bool isDir)
        {
            return true;
        } 
        
        private string _formEntityName(DirectoryInfo f)
        {
            StringBuilder sb = new StringBuilder();
            string name = f.Name;
            int nameLen = name.Length;
            if( nameLen > LCD_Manager.SCREEN_COLS - 9) nameLen = LCD_Manager.SCREEN_COLS - 9;
            sb.Append( name.Substring( 0, nameLen));
            while( sb.Length < LCD_Manager.SCREEN_COLS - 9) sb.Append(' ');
            sb.Append( " [DIR]");
            return sb.ToString();
        }

        private string _formEntityName(FileInfo f)
        {
            StringBuilder sb = new StringBuilder();
            string name = f.Name;
            int nameLen = name.Length;
            if (nameLen > LCD_Manager.SCREEN_COLS - 9) nameLen = LCD_Manager.SCREEN_COLS - 9;
            sb.Append(name.Substring(0, nameLen));
            while (sb.Length < LCD_Manager.SCREEN_COLS - 9) sb.Append(' ');
            sb.Append(_appendFileSize(f));
            return sb.ToString();
        }

        private string _appendFileSize(FileInfo f)
        {
            long size = f.Length;
            StringBuilder sb = new StringBuilder();
            sb.Append(' ');
            if (size < 8192L)
            {
                sb.Append( size.ToString("0000"));
                sb.Append('b');
                return sb.ToString();
            }
            size >>= 10;
            if (size < 8192L)
            {
                sb.Append(size.ToString("0000"));
                sb.Append('k');
                return sb.ToString();
            }
            size >>= 10;
            if (size < 8192L)
            {
                sb.Append(size.ToString("0000"));
                sb.Append('M');
                return sb.ToString();
            }
            size >>= 10;
            sb.Append(size.ToString("0000"));
            sb.Append('G');
            return sb.ToString();
        }
    }
}
